components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Storage/PublishedNodesProvider.cs

```diff
@@ -0,0 +1,159 @@
+
+            if (string.IsNullOrWhiteSpace(directory)) {
+                directory = Environment.CurrentDirectory;
+            }
+
+            var file = Path.GetFileName(_legacyCliModel.PublishedNodesFile);
+            FileSystemWatcher = new FileSystemWatcher(directory, file);
+
+            _lock = new SemaphoreSlim(1, 1);
+        }
+
+        /// <summary>
+        /// Get last write time of published nodes file.
+        /// </summary>
+        /// <returns></returns>
+        public DateTime GetLastWriteTime() {
+            _lock.Wait();
+            try {
+                return File.GetLastWriteTime(_legacyCliModel.PublishedNodesFile);
+            }
+            finally {
+                _lock.Release();
+            }
+        }
+
+        /// <summary>
+        /// Read content of published nodes file.
+        /// </summary>
+        /// <returns></returns>
+        public string ReadContent() {
+            _lock.Wait();
+            try {
+                using (var fileStream = new FileStream(
+                    _legacyCliModel.PublishedNodesFile,
+                    FileMode.Open,
+                    FileAccess.Read,
+                    FileShare.Read
+                 )) {
+                    return fileStream.ReadAsString(Encoding.UTF8);
+                }
+            }
+            catch (Exception e) {
+                _logger.Error(e, "Failed to read content of published nodes file from \"{path}\"",
+                    _legacyCliModel.PublishedNodesFile);
+                throw;
+            }
+            finally {
+                _lock.Release();
+            }
+
+        }
+
+        /// <summary>
+        /// Write new content to published nodes file
+        /// </summary>
+        /// <param name="content"> Content to be written. </param>
+        /// <param name="disableRaisingEvents"> If set FileSystemWatcher notifications will be disabled while updating the file.</param>
+        public void WriteContent(string content, bool disableRaisingEvents = false) {
+            _lock.Wait();
+            try {
+                // Store current state.
+                var eventState = FileSystemWatcher.EnableRaisingEvents;
+
+                if (disableRaisingEvents) {
+                    FileSystemWatcher.EnableRaisingEvents = false;
+                }
+
+                try {
+                    using (var fileStream = new FileStream(
+                        _legacyCliModel.PublishedNodesFile,
+                        FileMode.Open,
+                        FileAccess.Write,
+                        // We will require that there is no other process using the file.
+                        FileShare.None
+                     )) {
+                        fileStream.Write(Encoding.UTF8.GetBytes(content));
+                    }
+                }
+                catch (IOException e) {
+
+                    // We will fall back to writing with ReadWrite access.
```

Comment:
Let's log that there was a problem and we will try now with more relaxed share permissions.

Reply:
Added log message.